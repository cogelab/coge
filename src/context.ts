import * as path from "path";
import * as execa from "execa";
import {Environment, EnvironmentOptions, LookupOptions, PromptModule} from "coge-environment";
import {Context, Prompter} from "./types";
import {FileResolver} from "./resolvers/file";
import {FileLoader} from "./loders";

export interface DefaultContextOptions extends EnvironmentOptions {
  cwd?: string;
  lookup?: Partial<LookupOptions> | boolean;
  prompt?: PromptModule;
  exec?: (action: string, body: string | Buffer) => any;
}

const LogColors = () => ({
  shell: 'blue',
})

const resolver = new FileResolver('.coge.js', FileLoader);


export class DefaultContext implements Context {
  cwd: string;
  env: Environment;

  static async create();
  static async create(env: Environment);
  static async create(opts: DefaultContextOptions);
  static async create(env: Environment, opts: DefaultContextOptions);
  static async create(env?: Environment | DefaultContextOptions, opts?: DefaultContextOptions) {
    const context = new DefaultContext(env, opts);
    Object.assign(context, await resolver.resolve(context.cwd));
    return context;
  }

  protected constructor(env?: Environment | DefaultContextOptions, opts?: DefaultContextOptions) {
    if (!(env instanceof Environment)) {
      opts = env;
      env = undefined;
    }
    opts = opts || {};
    this.cwd = opts.cwd || process.cwd();
    if (opts.exec) this.exec = opts.exec;

    const {prompt} = opts;
    this.env = env || Environment.createEnv({prompt});

    Object.assign(this.logger.colors, LogColors());

    // lookup built-in templates
    this.env.lookup({localOnly: true, npmPaths: path.resolve(__dirname, '..', 'templates')});
    this.env.lookup({localOnly: true, packagePaths: path.resolve(__dirname, '..', 'templates')});

    // lookup local templates
    if (path.relative(this.cwd, path.resolve(__dirname, '..'))) {
      this.env.lookup({localOnly: true, npmPaths: path.resolve(this.cwd, 'templates')});
      this.env.lookup({localOnly: true, packagePaths: path.resolve(this.cwd, 'templates')});
    }

    // lookup template-* modules
    if (opts.lookup) {
      this.env.lookup(typeof opts.lookup === 'boolean' ? {} : opts.lookup);
    }

  }

  get logger() {
    return this.env.adapter.logger;
  }

  get prompter(): Prompter<any, any> {
    return this.env.adapter;
  }

  exec(action: string, body: string | Buffer) {
    const opts = body && body.length > 0 ? {input: body} : {}
    return execa(action, {shell: true, ...opts});
  }

  lookup(options?: Partial<LookupOptions> | boolean) {
    return this.env.lookup(options);
  }

}
