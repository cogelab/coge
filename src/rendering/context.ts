import inflection = require('inflection');
import changeCase = require('change-case');

import {Context} from "../types";
import {undasherize} from "../utils";

// supports kebab-case to KebabCase
// @ts-ignore
inflection.undasherize = undasherize;

const helpers = {
  capitalize(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();
  },
  inflection,
  changeCase
};

const doCapitalization = (hsh, [key, value]) => {
  hsh[key] = value;

  if (localsToCapitalize.includes(key)) hsh[helpers.capitalize(key)] = helpers.capitalize(value);

  return hsh;
};

const localsToCapitalize = ['name'];
const localsDefaults = {
  name: 'unnamed',
  group: '',
  folder: '',  // the folder of the template file
  dir: '',     // the relative directory to template root
  trd: '',     // the relative directory to template root
  drd: '',     // the relative directory to cwd
};

const capitalizedLocals = (locals: any) => Object.entries(locals).reduce(doCapitalization, {});

export const buildContext = (locals: any, context?: Context, ...extras: Record<string, any>[]) => {
  const localsWithDefaults = Object.assign({}, localsDefaults, locals, ...extras);
  const configHelpers = typeof context?.helpers === "function" ? context.helpers(locals, context) : (context?.helpers || {});
  return Object.assign(localsWithDefaults, capitalizedLocals(localsWithDefaults), {
    h: {...helpers, ...configHelpers}
  });
};
