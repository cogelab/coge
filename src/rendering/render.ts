import fs = require('fs-extra');
import ejs = require('ejs');
import fm = require('front-matter');
import path = require('path');
import walk = require('ignore-walk');

import {Context, RenderedAction, GeneratorEntry} from '../types'
import {buildContext} from './context';
import last from "@tiopkg/utils/array/last";

const ignores = [
  'coge.toml',
  'prompt.js',
  'index.js',
  '.cogeignore',
  '.DS_Store',
  '.Spotlight-V100',
  '.Trashes',
  'ehthumbs.db',
  'Thumbs.db',
];

function extractFolder(file: string) {
  const dir = path.dirname(file);
  const parts = dir.split(path.sep);
  return last(parts);
}

function renderTemplate(tmpl: any, locals?: Record<string, any>, context?: Context, extra?: Record<string, any>) {
  return typeof tmpl === 'string' ? ejs.render(tmpl, buildContext(locals, context, extra || {})) : tmpl;
}

async function listFiles(dir: string) {
  return walk
    .sync({path: dir, ignoreFiles: ['.cogeignore']})
    .map(f => path.join(dir, f));
}

export const render = async (
  context: Context,
  template: Pick<GeneratorEntry, 'dir' | 'pattern'>,
  locals?: Record<string, any>,
): Promise<RenderedAction[]> => {
  const files = (await listFiles(template.dir))
    .sort((a, b) => a.localeCompare(b))  // TODO: add a test to verify this sort
    .filter(file => !ignores.find(ig => file.endsWith(ig)))
    .filter(file => (template.pattern ? file.match(template.pattern) : true));

  const entries = await Promise.all(files.map(file => fs.readFile(file).then(text => ({file, text: text.toString()}))));
  return entries
    .map(({file, text}) => ({file, ...fm<Record<string, any>>(text)}))
    .map(({file, attributes, body}) => ({
      file,
      attributes: Object.entries(attributes).reduce((obj, [key, value]) => ({
        ...obj,
        [key]: renderTemplate(value, locals, context, {folder: extractFolder(file)}),
      }), {}),
      body: renderTemplate(body, locals, context, {folder: extractFolder(file)}),
    }));
}

