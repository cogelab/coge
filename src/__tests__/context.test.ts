import {buildContext} from '../rendering';
import {MockContext} from "./mocks/context";

describe('context', () => {
  it('should populate with capitalized keys', () => {
    expect(buildContext({name: 'foobar'})).toMatchSnapshot()
  })
  it('allows helpers to be initialized with current context', () => {
    const helpers = (args, context) => ({
      testArgs: () => args,
      testCtx: () => context,
    })
    const locals = {}
    const context = new MockContext({
      helpers,
    });
    const ctx = buildContext(locals, context)
    expect(typeof ctx.h.testArgs).toBe('function')
    expect(typeof ctx.h.testCtx).toBe('function')
    expect(ctx.h.testArgs()).toBe(locals)
    expect(ctx.h.testCtx()).toBe(context)
  })
})
