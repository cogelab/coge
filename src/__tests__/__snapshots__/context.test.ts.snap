// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context should populate with capitalized keys 1`] = `
Object {
  "Name": "Foobar",
  "group": "",
  "h": Object {
    "capitalize": [Function],
    "changeCase": Object {
      "camelCase": [Function],
      "camelCaseTransform": [Function],
      "camelCaseTransformMerge": [Function],
      "capitalCase": [Function],
      "capitalCaseTransform": [Function],
      "constantCase": [Function],
      "dotCase": [Function],
      "headerCase": [Function],
      "noCase": [Function],
      "paramCase": [Function],
      "pascalCase": [Function],
      "pascalCaseTransform": [Function],
      "pascalCaseTransformMerge": [Function],
      "pathCase": [Function],
      "sentenceCase": [Function],
      "sentenceCaseTransform": [Function],
      "snakeCase": [Function],
    },
    "inflection": Object {
      "_apply_rules": [Function],
      "camelize": [Function],
      "capitalize": [Function],
      "classify": [Function],
      "dasherize": [Function],
      "demodulize": [Function],
      "foreign_key": [Function],
      "humanize": [Function],
      "indexOf": [Function],
      "inflect": [Function],
      "ordinalize": [Function],
      "pluralize": [Function],
      "singularize": [Function],
      "tableize": [Function],
      "titleize": [Function],
      "transform": [Function],
      "undasherize": [Function],
      "underscore": [Function],
      "version": "1.12.0",
    },
  },
  "name": "foobar",
}
`;
